### PR TITLE
refactor!: use `rsa` instead of `openssl`

### PR DESCRIPTION
**Pull request by kwaa**
closed https://github.com/LemmyNet/activitypub-federation-rust/issues/49

cc @Nutomic 

This PR uses `rsa` crate instead of `openssl` to make the project more portable.

The private_key [here](https://github.com/LemmyNet/activitypub-federation-rust/pull/116/files#diff-e43e9f7c705fd8ae331a3d9ced2894b39cc21b2c7a4f46d612b150ab55ef762eR380) looks isn't in pkcs8 format, so I've used `from_pkcs1_pem` here.

### CODE REVIEW

**Comment by Nutomic** [comment]
Very nice! To fix the test failure you need to merge commits from main branch to yours.

**Comment by kwaa** [comment]
> To fix the test failure you need to merge commits from main branch to yours.

The master branch doesn't seem to have any new commits. https://github.com/LemmyNet/activitypub-federation-rust/commits/main/

**Comment by Nutomic** [comment]
My bad, I didnt commit/push my change :facepalm: 

https://github.com/LemmyNet/activitypub-federation-rust/pull/117

File: Cargo.toml [hunk 1]
```diff
@@ -41,10 +41,11 @@ reqwest = { version = "0.11.27", features = ["json", "stream"] }
 reqwest-middleware = "0.2.5"
 tracing = "0.1.40"
 base64 = "0.22.1"
-openssl = "0.10.64"
+rand = "0.8.5"
+rsa = "0.9.6"
 once_cell = "1.19.0"
 http = "0.2.12"
-sha2 = "0.10.8"
+sha2 = { version = "0.10.8", features = ["oid"] }
```
Comment (Nutomic) [inline Cargo.toml:48]: Whats this feature needed for?

Comment (kwaa) [reply Cargo.toml:48]: > Whats this feature needed for?

`Pkcs1v15Sign::new::<Sha256>()`  needs this feature.

File: Cargo.toml [hunk 2]
```diff
@@ -82,11 +88,10 @@ axum = { version = "0.6.20", features = [
...
 hyper = { version = "0.14", optional = true }
-http-body-util = {version = "0.1.1", optional = true }
+http-body-util = { version = "0.1.1", optional = true }
 
 [dev-dependencies]
 anyhow = "1.0.82"
-rand = "0.8.5"
 env_logger = "0.11.3"
 tower-http = { version = "0.5.2", features = ["map-request-body", "util"] }
 axum = { version = "0.6.20", features = [
```
[the remark this answers is not in the document]
Comment (kwaa) [reply Cargo.toml:97]: > You might as well add the toml formatting to CI. You can [copy it from the Lemmy repo](https://github.com/LemmyNet/lemmy/blob/main/.woodpecker.yml#L57-L62).

I'll take a look, but maybe changing CI as another PR would be better.

**Review by Nutomic — APPROVED** [review]
Ive confirmed that Lemmy federation tests are passing with these changes. So you just need to pull in the test fix in order to merge.

**Comment by kwaa** [comment]
> Ive confirmed that Lemmy federation tests are passing with these changes. So you just need to pull in the test fix in order to merge.

I updated the branch.

**Comment by dessalines** [comment]
This should also probably be done for lemmy too. I'll open an issue.

**Pull request merged**